Bluetooth: host: crypto: Add a config to select how random is generated

LingaoM opened the pull request:
Considering that in most scenarios, `bt_rand` will not be called frequently, 
but the current implementation of tinycrypt will occupy more than 300 bytes
of RAM space. Its existence is to optimize the frequent call of `bt_rand`. 

Therefore, it is considered to put it into a config(`BT_HOST_CRYPTO_RANDOM`),
when this config has been selected, will use tinycrypt library for random.
Otherwise will call bt random hci command.


Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

carlescufi commented:
@jhedberg @cvinayak do you remember why we have this rand implementation in the Host at all? Is it to support controllers that do not implement the LE_RAND command?

LingaoM commented on subsys/bluetooth/host/crypto.c:
```diff
@@ -27,92 +27,41 @@
-	size_t i;
-	int ret;
-
-	for (i = 0; i < (sizeof(seed) / 8); i++) {
-		struct bt_hci_rp_le_rand *rp;
-		struct net_buf *rsp;
-
-		ret = bt_hci_cmd_send_sync(BT_HCI_OP_LE_RAND, NULL, &rsp);
-		if (ret) {
-			return ret;
-		}
-
-		rp = (void *)rsp->data;
-		memcpy(&seed[i * 8], rp->rand, 8);
-
-		net_buf_unref(rsp);
-	}
-
-	extra = k_uptime_get();
-
-	ret = tc_hmac_prng_reseed(h, seed, sizeof(seed), (uint8_t *)&extra,
-				  sizeof(extra));
-	if (ret == TC_CRYPTO_FAIL) {
-		BT_ERR("Failed to re-seed PRNG");
-		return -EIO;
-	}
-
-	return 0;
-}
-
 int prng_init(void)
 {
-	struct bt_hci_rp_le_rand *rp;
-	struct net_buf *rsp;
-	int ret;
-
 	/* Check first that HCI_LE_Rand is supported */
 	if (!BT_CMD_TEST(bt_dev.supported_commands, 27, 7)) {
 		return -ENOTSUP;
 	}
```
@carlescufi I'm confused with this.

cvinayak reviewed (dismissed):
Original implementation uses fast pseudo random numbers which has been initialized by a true random number as seed.

Purely using random numbers over HCI interface will be very slow/time consuming and delay the caller that is using the random numbers.

LingaoM commented:
I need to explain first that more than 300 bytes of ram is used for simple random use, I don’t think it’s worth it, I simply browse the length used to call random, which are basically 16 bytes, that is, only need two hci transfer.

LingaoM commented:
Is it possible to add a config to choose which method to use, fast but with large resource overhead or slow with low resource overhead.

LingaoM committed:
Bluetooth: host: crypto: Add a config to select how rand is generated

Considering that in most scenarios, bt_rand will not be called
frequently, but the current implementation of tinycrypt will
occupy more than 300 bytes of RAM space. Its existence is to
optimize the frequent call of bt_rand.

Therefore, it is considered to put it into a config
(`BT_HOST_CRYPTO_RANDOM`), when this config has been selected,
will use tinycrypt library for random. Otherwise will call bt random
hci command.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

Thalley commented on subsys/bluetooth/host/crypto.c:
```diff
@@ -91,9 +92,13 @@ int prng_init(void)
+	ret = prng_reseed(&prng);
+#endif /*CONFIG_BT_HOST_CRYPTO_RANDOM*/
+
+	return ret;
```
The change here is to simply return 0 if `CONFIG_BT_HOST_CRYPTO_RANDOM=n`. Would it make more sense to just guard the entire function by `CONFIG_BT_HOST_CRYPTO_RANDOM`, so that it isn't callable?

LingaoM replied:
We need to check whether the controller supports the BT RANDOM command.

Thalley replied:
Why? 

if `!IS_ENABLED(CONFIG_BT_HOST_CRYPTO_RANDOM` then it doesn't really matter if the controller supports it or not, right? Because it will just return 0 in any case, and not perform any actions.

LingaoM replied:
Add `BT_CMD_TEST(bt_dev.supported_commands, 27, 7)` check. I don't think it's unnecessary. After all, in the later operation, the bt random command will be issued directly https://github.com/zephyrproject-rtos/zephyr/blob/69829d2a5698c7621c1006bbec2d953017b56e29/subsys/bluetooth/host/crypto.c#L132

LingaoM replied:
Without this check, it is possible to issue a command that is not supported.

Thalley replied:
Not sure I've expressed myself precisely. 

If `CONFIG_BT_HOST_CRYPTO_RANDOM=n` then the function does nothing, except return "success", unless the controller does not support it. 

Why do we need to check controller support if we are not going to use the controller?

Thalley commented on subsys/bluetooth/host/crypto.c:
```diff
@@ -114,6 +119,34 @@ int bt_rand(void *buf, size_t len)
+#else /* !CONFIG_BT_HOST_CRYPTO_RANDOM */
+int bt_rand(void *buf, size_t len)
+{
+	int ret, size, i = 0;
+
+	while (len) {
+		struct bt_hci_rp_le_rand *rp;
+		struct net_buf *rsp;
+
```
```suggestion
	size_t i = 0;

	while (len) {
		struct bt_hci_rp_le_rand *rp;
		struct net_buf *rsp;
		size_t size;
		int ret;

```

Thalley commented on subsys/bluetooth/host/crypto.c:
```diff
@@ -114,6 +119,34 @@ int bt_rand(void *buf, size_t len)
+		size = MIN(len, 8);
+
+		rp = (void *)rsp->data;
+		memcpy((uint8_t *)buf + i, rp->rand, size);
```
```suggestion
		(void)memcpy((uint8_t *)buf + i, rp->rand, size);
```

LingaoM commented on subsys/bluetooth/host/crypto.c:
```diff
@@ -114,6 +119,35 @@ int bt_rand(void *buf, size_t len)
+		struct bt_hci_rp_le_rand *rp;
+		struct net_buf *rsp;
+
+		ret = bt_hci_cmd_send_sync(BT_HCI_OP_LE_RAND, NULL, &rsp);
+		if (ret) {
+			return ret;
+		}
```
@Thalley We use controller (https://github.com/zephyrproject-rtos/zephyr/pull/42264#discussion_r814699937)

LingaoM replied:
This scheme is to directly call the BT Rand HCI command, while previously, by using the BT Rand HCI command as a random seed, both will call the controller.

Thalley replied:
Right, but why are doing the check for supporting `HCI_LE_Rand` in a function that doesn't call `bt_hci_cmd_send_sync(BT_HCI_OP_LE_RAND, NULL, &rsp);`, rather than doing it here? 

What I suggest is to guard the entire `prng_init` function with `CONFIG_BT_HOST_CRYPTO_RANDOM` and then add the 
```c
if (!BT_CMD_TEST(bt_dev.supported_commands, 27, 7)) {
		return -ENOTSUP;
}
```
check here as well. 

Using `prng_init` just for a controller check seem like a poor design to me.

LingaoM replied:
It's not friendly in my opinion to have to make judgments before each call. During initialization. Judging once is the easiest way.

Thalley replied:
Judging before each call is how most other HCI calls are implemented. 

The current implementation of `bt_rand` when `CONFIG_BT_HOST_CRYPTO_RANDOM=y` tests for `if (ret == TC_HMAC_PRNG_RESEED_REQ)` which is implicitly a check for `BT_CMD_TEST(bt_dev.supported_commands, 27, 7)`, so we already check each time now. 

Furthermore, the only place where `prng_init` is called is in `common_init` where the call is guarded by `IS_ENABLED(CONFIG_BT_HOST_CRYPTO)`, so there's absolutely no reason not to guard `prng_init` with CONFIG_BT_HOST_CRYPTO. For that same reason, the check for `BT_CMD_TEST(bt_dev.supported_commands, 27, 7)` in `prng_init` is, in this PRs current state, never done when `CONFIG_BT_HOST_CRYPTO_RANDOM=n`

Thalley replied:
CC?

LingaoM replied:
I want to try to avoid the situation that the controller does not support bt rand, but the host initiates this command.

LingaoM replied:
Anyway, I have made changes, but I still want to state that BT_HOST_CRYPTO_RANDOM depends on BT_HOST_CRYPTO.

Even if BT_HOST_CRYPTO_RANDOM=y or =n, it still makes sense for the initialization function prng_init to exist.

LingaoM replied:
> if (ret == TC_HMAC_PRNG_RESEED_REQ) which is implicitly a check for BT_CMD_TEST(bt_dev.supported_commands, 27, 7), so we already check each time now.

Still trying to understand is there aimplicitly a check here, I don't understand.

`TC_HMAC_PRNG_RESEED_REQ` is tinycrypt, nothing to do with bluetooth.

Thalley replied:
> > if (ret == TC_HMAC_PRNG_RESEED_REQ) which is implicitly a check for BT_CMD_TEST(bt_dev.supported_commands, 27, 7), so we already check each time now.
> 
> Still trying to understand is there implicitly a check here, I don't understand.
> 
> `TC_HMAC_PRNG_RESEED_REQ` is tinycrypt, nothing to do with bluetooth.

The host implementation of `bt_rand` calls `tc_hmac_prng_generate` and handles `TC_HMAC_PRNG_RESEED_REQ` so it is related to Bluetooth. 

I still stand by that calling `prng_init` only as a check for `if (!BT_CMD_TEST(bt_dev.supported_commands, 27, 7))` before calling calling `bt_rand`, instead of moving the check for `if (!BT_CMD_TEST(bt_dev.supported_commands, 27, 7))` into `bt_rand` is a poor design. 

Not only is `prng_init` not a public function (and thus the application can never get that result), but it is a function that is entirely unrelated to calling `bt_rand` when using the HCI command to generate a random value. 

In the case that an application tries to call `bt_rand`, and the controller does not support it, this will just result in a `-EIO` return code, which will be difficult for the application to debug. 

So I still suggest guarding `prng_init` with `CONFIG_BT_HOST_CRYPTO_RANDOM` and moving the check for controller support to `bt_rand` if `CONFIG_BT_HOST_CRYPTO_RANDOM=n`

LingaoM replied:
This time it seems to convince me.

Thalley replied:
My apologies for being unclear :)

LingaoM commented:
@Thalley The previous commitments information is too simple and may not be easy for you to understand. I revised it again. I'm very sorry.

LingaoM commented:
Any Problem?

Thalley commented on subsys/bluetooth/host/crypto.c:
```diff
@@ -114,6 +115,40 @@ int bt_rand(void *buf, size_t len)
+			return ret;
+		}
+
+		size = MIN(len, 8);
```
Please avoid magic numbers

```suggestion
		size = MIN(len, sizeof(rsp->rand));
```

Thalley commented on subsys/bluetooth/host/crypto.c:
```diff
@@ -114,6 +115,40 @@ int bt_rand(void *buf, size_t len)
+
+		size = MIN(len, 8);
+
+		rp = (void *)rsp->data;
+		(void)memcpy((uint8_t *)buf + i, rp->rand, size);
```
Should we check `rp->status` before copying the data?

Thalley commented:
> Any Problem?

Please be patient with reviewers :)

Thalley commented on subsys/bluetooth/host/crypto.c:
```diff
@@ -114,6 +115,44 @@ int bt_rand(void *buf, size_t len)
+			return ret;
+		}
+
+		size = MIN(len, sizeof(rp->rand));
+
+		rp = (void *)rsp->data;
+		if (rp->status) {
+			return -EIO;
+		}
+
+		(void)memcpy((uint8_t *)buf + i, rp->rand, size);
```
```suggestion
		rp = (void *)rsp->data;
		if (rp->status) {
			return -EIO;
		}

		size = MIN(len, sizeof(rp->rand));
		(void)memcpy((uint8_t *)buf + i, rp->rand, size);
```

Minor comment: We don't need to calculate size if rp->status is an error

LingaoM replied:
Thank you for your careful consideration.

jhedberg commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -109,16 +109,29 @@ config BT_DRIVER_RX_HIGH_PRIO
+	  This will consume additional ram, but may speed up the generation of random
+	  numbers.
+
+	  Otherwise, random numbers will be generated through multiple HCI calls,
+	  which will not consume additional resources, but may take a long time,
+	  depending on the length of the random.
```
There's a missing noun at the end of the sentence ("random" is an adjective). Probably it should end with "length of the random data".

jhedberg commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -109,16 +109,29 @@ config BT_DRIVER_RX_HIGH_PRIO
 	select TINYCRYPT
 	select TINYCRYPT_AES
+
+config BT_HOST_CRYPTO_RANDOM
```
Can we call this `BT_HOST_CRYPTO_PRNG`, since that's really what it's affecting: when enabled we use a PRNG to generate random data, and when disabled we rely on HCI_LE_Rand.

joerchan commented on subsys/bluetooth/host/crypto.c:
```diff
@@ -114,6 +115,44 @@ int bt_rand(void *buf, size_t len)
+		return -ENOTSUP;
+	}
+
+	while (len) {
```
This code duplicates the implementation of the entropy over HCI source:
https://github.com/zephyrproject-rtos/zephyr/pull/41424/files#diff-cd0469090601a75fd60f22e823813f55b66de8d019fd2203a33cb10b5bcee433R31

Wouldn't it make sense to combine bt_rand and the entropy over HCI into a single entity?

In a combined build using the entropy source for HCI is now less efficient than calling bt_rand, because bt_rand has the controller/crypto.c implementation.

To me it seems that bt_rand is still the best option for the application to use, but entropy source is the most portable option.
In my opinion it would make more sense to either:
- Entropy source over bt_rand instead of HCI.
- Deprecate bt_rand and use the entropy source in the host instead.

carlescufi replied:
cc @JordanYates

carlescufi replied:
Please replace this with the use of the `bt-hci` entropy driver:

```
entropy_driver = DEVICE_DT_GET(DT_CHOSEN(zephyr_entropy));
rc = entropy_get_entropy(entropy_driver, (uint8_t *)&state, sizeof(state));
```

joerchan commented:
@LingaoM What is the performance difference between the options?
If it is very small then maybe it doesn't even make sense to keep the tinycrypt behavior.

LingaoM commented:
> @LingaoM What is the performance difference between the options? If it is very small then maybe it doesn't even make sense to keep the tinycrypt behavior.

I didn't think of a reasonable test method. Maybe others can try it.
